src/qibocal/auto/task.py

```diff
@@ -86,6 +87,21 @@ def operation(self):
 
         return Operation[self.action.operation].value
 
+    @property
+    def validator(self) -> Validator:
+        """Validator object from ValidationSchemes Enum"""
+        if self.action.validator is None:
+            return None
+        return ValidationSchemes[self.action.validator["scheme"]].value.load(
+            self.action.validator
+        )
```

Comment:
If the `scheme` attribute is required, make `validator` already a `dataclass`, and deserialize `scheme` in it.

Moreover, this procedure could also be offloaded to the object in which `self.action.validator` is deserialized into, and leave here
```suggestion
        return self.action.validator.load()
```

Reply:
Notice that you already have such an object, is exactly the `Validator` class - I'm just suggesting to make it a `pydantic.dataclasses.dataclass` and deserialize in it the possible parameters as well.

Actually, you would not even need to `load` in that case, and you would not even need the `property` at all, just return the `Validator` instance created during the runcard deserialization.

Reply:
Good point. In fact I called it `validator` before and `pydantic` complained saying that they have already a validator...

Reply:
One quick question. I tried to implement the deserialization in `Runcard` and it almost works....
The only problem is that I would like to deserialize depending on the `scheme`. For example in this case if I pass `scheme: chi2` I would like to deserialize using `chi2` in `ValidationSchemes` which will be mapped to `Chi2Validator`. Right now I am only able to deserialize to the abstract class `Validator`, which I believe it is what I should put in `Runcard`.

https://github.com/qiboteam/qibocal/blob/dcc5abdc5108cad4ecf5df01f481bfb0e11df484/src/qibocal/auto/validation.py#L61-L64

I was looking for a way to deserialize like this on poetry but I could not find it.
Have you already encountered this problem @AleCandido?

Reply:
I guess with "Poetry" you meant "Pydantic". I believe there are two ways of doing it:

1. attempt to deserialize a `Union`
2. implement custom deserialization in `Validator`, e.g. using a [pre/post hook](https://docs.pydantic.dev/latest/concepts/dataclasses/#initialization-hooks) (or a combination of the two)

I see problems in both, in the sense that no one is a perfect solution out of the box. 
The `Union` would work perfectly if you do not allow for arbitrary parameters, and distinguish the various validators based on their parameters (rather than based on a tag). But this is not always sensible, since you might want to have different validators logic with the same parameters.
The custom deserialization is slightly annoying, because you have to write more custom logic, and it seems there are no hooks in which you get the parameters and output the object (at least for dataclasses), so you should first let Pydantic construct the `Validator` object (base class), and then create a new object of the given type in the post hook.

The problem is that, in this case, your type is intrinsically a `Union` (not an `Enum`, that is a single type, but different values, well [supported](https://docs.pydantic.dev/latest/api/standard_library_types/#enum) by Pydantic), but since you're discriminating them on a value (not the key, but the value in the dictionary) this should be done *at runtime* (so it can't be inferred from the type).

Reply:
Actually, an alternative would be to give up on the different subclasses, and have a single class `Validator`, while the logic is delegated to functions (so you would only pick the function dynamically, and Pydantic can safely construct the `Validator` object, that will later look up for the function).

Same with classes, of course, if you give up on subclassing, and use composition instead (so you will later load in the validator the validation logic as an attribute).
But if functions are already working, they are just simpler.

Reply:
> I guess with "Poetry" you meant "Pydantic".

Yes, indeed I meant "Pydantic"...

Thanks for detailed summary.

> Actually, an alternative would be to give up on the different subclasses, and have a single class `Validator`, while the logic is delegated to functions (so you would only pick the function dynamically, and Pydantic can safely construct the `Validator` object, that will later look up for the function).

I guess this will be the easiest option. I don't really feel like spending too much time trying to implement 1. or 2.
Eventually if we see that deserializing with Pydantic is too complicated we can always fall back to the present solution already implemented in this PR.

Reply:
I manage to implement the approach you suggested. Here is the updated runcard:

```yaml
platform: dummy

qubits: [0,1]


actions:

  - id: ramsey
    priority: 0
    operation: ramsey
    validator:
      scheme: chi2
      parameters:
        chi2_max_value: 0.1
    parameters:
      delay_between_pulses_start: 16 # must be a multiple of 4 incl 0
      delay_between_pulses_end: 1000
      delay_between_pulses_step: 4 # must be a multiple of 4
      n_osc: 10
```

Reply:
It looks great :D

Reply:
Also this validator now only works for results. With this we should be able to also have validators for data directly. For example we could test whether we are seeing only noise (for example using some SNR).
The missing step would be to define some sort of handler, but we can keep it for another PR.

Reply:
> With this approach we should be able to also have validators for data directly.

In principle yes, but this could be done even during the `fit` call, if relevant for the routine, so (for the time being) I would avoid it.
It is true that we could make the single routines more composable, creating some building blocks that could be added (as the SNR). But keep it on the side, and let's try to converge on the exceptional flow first.